### PR TITLE
Use internal buffer in RobotLogger

### PR DESCRIPTION
**Pull request by luator**
[//]: # "Thanks for your contribution.  To make life of the reviewers easier,"
[//]: # "please give this pull request a meaningful title and provide the"
[//]: # "requested information below."

## Description

Refactor the `RobotLogger` and add a logging mode that uses an internal buffer (same as in SensorLogger).  This should resolve some issues we had in the past where some time steps were missing in the log because they where not in the robot data time series anymore.

This new logging method supports writing to either binary, CSV or gzipped CSV files.

The old `start()`/`stop()` methods (used for continuous writing to file) have been renamed to `start|stop_continuous_writing()`.  Further this methods are marked deprecated as I plan to remove them soon, to not make the logger too bloated (this logging method is anyway not used at the moment).

To avoid confusion with the internal buffer, rename the `write_current_buffer*` methods to `save_current_robot_data*`.  The old names are kept as aliases for now but are marked deprecated.

### Other Changes
- Use applied_action in RobotFrontend::wait_for_timeindex.  The observation is the first time series to be filled by the backend, so when waiting only for this, it is not guaranteed that the other time series have already been filled.  To ensure this, use the applied_action time series instead, which is the last one to be filled.
- Avoid blocking in SensorLogger in case the backend has stopped.

## How I Tested

By running the unit tests.

## I fulfilled the following requirements

[//]: # "Please make sure you followed these steps before requesting a review."
[//]: # "Check the boxes in the list below, when done."

- [x] All new code is formatted according to our style guide (for C++ run clang-format, for Python, run flake8 and fix all warnings).
- [x] All new functions/classes are documented and existing documentation is updated according to changes.
- [x] No commented code from testing/debugging is kept (unless there is a good reason to keep it).

### CODE REVIEW

**Comment by luator** [comment]
Rebased and adjusted to changes on master.

File: tests/test_robot_logger.cpp [hunk 1]
```diff
@@ -54,21 +60,84 @@ class TestRobotLogger : public ::testing::Test
...
+
+        // check data
+        for (uint32_t i = start_t; i < end_t; i++)
+        {
+            int time_index;
+            double time_stamp, action_repetitions, error_status, obs_0, obs_1,
+                applied_0, applied_1, desired_0, desired_1;
+
+            ASSERT_TRUE(std::getline(infile, line))
+                << "Failed to read line " << i;
```
Comment (vincentberenz) [inline tests/test_robot_logger.cpp:122]: I am bit confused: in which stream this is writing ?

**Review by vincentberenz — APPROVED** [review]
lgtm

File: tests/test_robot_logger.cpp (hunk 1 above)
Comment (luator) [reply tests/test_robot_logger.cpp:122]: This is the error message that is displayed if the assert fails.

**Pull request merged**